noxfile.py

```diff
@@ -70,30 +70,95 @@ def docs(session: nox.Session):
 
 
 @nox.session
-def lint(session: nox.Session):
+def lint(session: nox.Session) -> None:
     """Check all files for linting errors"""
     session.run_always("pdm", "install", "-G", "tools", external=True)
 
     session.run("pre-commit", "run", "--all-files", *session.posargs)
 
 
 @nox.session(name="check-manifest")
-def check_manifest(session: nox.Session):
+def check_manifest(session: nox.Session) -> None:
     """Run check-manifest."""
     # --no-self is provided here because check-manifest builds disnake. There's no reason to build twice, so we don't.
     session.run_always("pdm", "install", "--no-self", "-dG", "tools", external=True)
     session.run("check-manifest", "-v")
 
 
 @nox.session()
-def slotscheck(session: nox.Session):
+def slotscheck(session: nox.Session) -> None:
     """Run slotscheck."""
     session.run_always("pdm", "install", "-dG", "tools", external=True)
     session.run("python", "-m", "slotscheck", "--verbose", "-m", "disnake")
 
 
+@nox.session
+def autotyping(session: nox.Session) -> None:
+    """Run autotyping.
+
+    Because of the nature of changes that autotyping makes, and the goal design of examples,
+    this runs on each folder in the repository with specific settings.
+    """
+    session.run_always("pdm", "install", "-dG", "codemod", external=True)
+
+    env = {"LIBCST_PARSER_TYPE": "native"}
+    base_command = ["python", "-m", "libcst.tool", "codemod", "autotyping.AutotypeCommand"]
+    commands = {
+        "disnake": [
+            "--aggressive",
+        ],
+        "examples": [
+            "--scalar-return",
+            "--bool-param",
+            "--bool-param",
+            "--int-param",
+            "--float-param",
+            "--str-param",
+            "--bytes-param",
+        ],
+        "scripts": [
+            "--aggressive",
+        ],
+        "tests": [
+            "--aggressive",
+        ],
+        "test_bot": [
+            "--aggressive",
+        ],
+        # also autotype this file
+        "noxfile.py": [
+            "--aggressive",
+        ],
+    }
+
+    if session.posargs:
+        # short circuit with the provided arguments
+        # if there's just one file argument, give it the defaults that we normally use
+        posargs = session.posargs
+        if len(posargs) == 1 and not posargs[0].startswith("--"):
+            module = posargs[0].split("/")[0]
+            try:
+                posargs += commands[module]
+            except KeyError:
+                pass
+        session.run(*base_command, *posargs, env=env)
+        return
+
+    # run the custom fixers
+    any_change_made = False
+    for module, options in commands.items():
+        try:
+            session.run(*base_command, module, *options, env=env)
+        except Exception as e:
+            session.log(e)
+            any_change_made = True
+
+    if any_change_made:
+        session.error("Typings were added")
```

Comment:
What's the reasoning for making this command error when changes were made? This is already taken into account when run in CI, and not all that relevant when running locally, imo.

Reply:
local development to know a change was made. Its mostly UX rather than anything else.

Reply:
https://github.com/DisnakeDev/disnake/pull/958/commits/ba338d7096c6077c34dfc2982bc894b9342ea5a5